Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp

```diff
@@ -80,6 +80,44 @@ NetworkStorageSession::~NetworkStorageSession()
     g_signal_handlers_disconnect_matched(m_cookieStorage.get(), G_SIGNAL_MATCH_DATA, 0, 0, nullptr, nullptr, this);
 }
 
+#if HAVE(COOKIE_CHANGE_LISTENER_API)
+void NetworkStorageSession::notifyCookieAdded(SoupCookie* newCookie)
+{
+    if (soup_cookie_get_http_only(newCookie))
+        return;
+
+    auto host = String::fromUTF8(soup_cookie_get_domain(newCookie));
+    for (auto& host : m_cookieChangeObservers.keys()) {
+        if (soup_cookie_domain_matches(newCookie, host.utf8().data())) {
+            auto observers = m_cookieChangeObservers.getOptional(host);
+            if (!observers)
+                break;
+
+            for (auto& observer : *observers)
+                observer.cookiesAdded(host, { Cookie(newCookie) });
+        }
+    }
+}
+
+void NetworkStorageSession::notifyCookieDeleted(SoupCookie* oldCookie)
+{
+    if (soup_cookie_get_http_only(oldCookie))
+        return;
+
+    auto host = String::fromUTF8(soup_cookie_get_domain(oldCookie));
+    for (auto& host : m_cookieChangeObservers.keys()) {
+        if (soup_cookie_domain_matches(oldCookie, host.utf8().data())) {
+            auto observers = m_cookieChangeObservers.getOptional(host);
+            if (!observers)
+                break;
+
+            for (auto& observer : *observers)
+                observer.cookiesDeleted(host, { Cookie(oldCookie) });
```

Comment:
seeing the code is the same in both cases I wonder if it makes sense to refactor into a method that takes a lambda (which would invoke cookiesAdded or cookiesDeleted appropriately)

Reply:
This is not simpler.

Reply:
Fair enough, thanks for trying!